addons/pos_coupon/models/coupon_program.py

```diff
@@ -0,0 +1,84 @@
+
+# NOTE Use black to automatically format this code.
+
+from odoo import api, fields, models, _
+
+import ast
+
+class CouponProgram(models.Model):
+    _inherit = "coupon.program"
+
+    pos_config_ids = fields.Many2many(
+        "pos.config",
+        string="Point of Sales",
+        readonly=True,
+    )
+    pos_order_line_ids = fields.One2many(
+        "pos.order.line",
+        "program_id",
+        string="PoS Order Lines",
+        help="Order lines where this program is applied.",
+    )
+    promo_barcode = fields.Char(
+        "Barcode",
+        default=lambda self: self.env["coupon.coupon"]._generate_code(),
+        help="A technical field used as an alternative to the promo_code. "
+        "This is automatically generated when promo_code is changed.",
+    )
+    pos_order_ids = fields.Many2many(
+        "pos.order", help="The PoS orders where this program is applied.",
+    )
+    pos_order_count = fields.Integer(
+        "PoS Order Count", compute="_compute_pos_order_count"
+    )
+    valid_product_ids = fields.Many2many(
+        "product.product",
+        "Valid Products",
+        compute="_compute_valid_product_ids",
+        help="These are the products that are valid in this program.",
+    )
+    valid_partner_ids = fields.Many2many(
+        "res.partner",
+        "Valid Partners",
+        compute="_compute_valid_partner_ids",
+        help="These are the partners that can avail this program.",
+    )
+
+    @api.depends("pos_order_ids")
+    def _compute_pos_order_count(self):
+        for program in self:
+            program.pos_order_count = len(program.pos_order_ids)
+
+    def write(self, vals):
+        if "promo_code" in vals:
+            vals.update({"promo_barcode": self.env["coupon.coupon"]._generate_code()})
+        return super(CouponProgram, self).write(vals)
+
+    def action_view_pos_orders(self):
+        self.ensure_one()
+        return {
+            "name": _("PoS Orders"),
+            "view_mode": "tree,form",
+            "res_model": "pos.order",
+            "type": "ir.actions.act_window",
+            "domain": [("id", "in", self.pos_order_ids.ids)],
+            "context": dict(self._context, create=False),
+        }
+
+    @api.depends("rule_products_domain")
+    def _compute_valid_product_ids(self):
+        for program in self:
+            domain = ast.literal_eval(program.rule_products_domain) if program.rule_products_domain else []
+            program.valid_product_ids = self.env["product.product"].search(domain).ids
+
+    @api.depends("rule_partners_domain")
+    def _compute_valid_partner_ids(self):
+        for program in self:
+            domain = ast.literal_eval(program.rule_partners_domain) if program.rule_partners_domain else []
+            program.valid_partner_ids = self.env["res.partner"].search(domain).ids
+
+    def get_number_usage(self):
+        res = super(CouponProgram, self).get_number_usage()
+        return res + self.pos_order_count
```

Comment:
I see that "order_count" is still used in sale_coupon, shouldn't we use the sum of order_count and pos_order_count instead? Maybe we should have a computed field, total_order_count, that should be used in sale_coupon. And you can replace the rpc to get_number_usage bay a read of the new field on the coupon program.

Reply:
I get what you meant. I'll try that.

Reply:
can't you create a computed field instead of a function?

Reply:
I didn't add the `total_order_count` field. Instead, I replace the `order_count` from `maximum_number_use-vs-order_count` comparisons using `get_total_order_count()`. Moreover, I kept the rpc because its value is quite dynamic - it changes everytime an order is added.

Reply:
I was trying to create a computed field but when I'm at the part where we need to check maximum_number_use in pos_coupon, I realized that I can't save the value of total_order_count in the program because it's relatively dynamic. So even if I replace the get_total_order_count with a computed field, I will still need to perform a `search_read` rpc getting it's value. I just opted with less changes. But if you insist, I'll do the computed field.

Reply:
ok, but it si more logic to create a field computed.

Just create the field in coupon, and override the compute methods and adding depends in sale_coupon and pos_coupon.

keep the rpc, but call method 'read', with program_id and a list containing field name ['total_order_count']

Reply:
no need to store the compute field

Reply:
Got it.

Reply:
it is also because if we want to display the value in the form view of program to know how many remaining it left, it'll be easier

Reply:
I didn't think of that. I'm sold. :)